### PR TITLE
2.0 UI Overhaul/Refresh

### LINKED ISSUE #10953: Refresh Airflow UI
**Description**

As the Airflow 2.0 is coming nearer it would be awesome to **refresh** the Airflow WebUI. By this it mostly means:
- using some "modern colors" (how about Airflow logo palette?)
- fixing some padding/spacing
- and other small HTML / CSS touches 

**Use case / motivation**

The main point is to **refresh** the look, not to rewrite the Web UI.

**Related Issues**

Apache Airflow brand book: https://cwiki.apache.org/confluence/display/AIRFLOW/Brandbook

I'm far away from being a frontend expert but removing the "Airflow green" gives a good impression I think:

Screenshot 2020-09-14 at 10 45 22

### PR DESCRIPTION
**Pull request by ryanahamilton**
Resolves #10953.

A refreshed UI for the 2.0 release. The existing "theming" is a bit long in the tooth and this PR attempts to give it a modern look and some freshness to compliment all of the new features under the hood.

The majority of the changes to UI have been done through updates to the Bootstrap theme contained in `bootstrap-theme.css`. These are simply overrides to the default stylings that are packaged with Bootstrap.

I've tweaked, refactored, and cleaned up syntax across most of the view, script, and CSS files.

I've highlighted outstanding questions in **bold** below.

## Color

Implements a full refresh of the color palette based on the colors in the [brand book](https://cwiki.apache.org/confluence/display/AIRFLOW/Brandbook) (outlined in black below). Those colors have been expanded to a full color system robust enough to cover all aspects of the UI.

**Where would the most ideal place to document this?** Out of scope for this PR, but I would typically put these all into CSS variables in a shared import file as documentation.

image

## Iconography

The current icon system (Bootstrap Glyficons) has been replaced with the [Material Design Icons](https://material.io/resources/icons/) (Apache license version 2.0). They have a clean, modern look, and a robust selection. Like Glyficons, It is implemented via webfont. ~I attempted (and would have preferred) to install it via NPM package, but it appears that it is not kept up-to-date with their CDN/downloadable libraries, so instead am serving it via static files.~ The font has been converted to a Base64 encoding and embedded within the new `material-icons.css` stylesheet. This yields a cacheable file that doesn't require multiple requests (of importing font files).

Most of the icons are very similar to their predecessors, others I've attempted to select the best representation possible based on the available set.
Image 2020-09-29 at 2 32 16 PM

## Typography

The UI currently does not utilize either of the font-families (Rubik, Operator Mono) designated by the brand guidelines. I did some experimentation with these, but there were some immediate, glaring problems in regards to the overall widths of Rubik. I may follow-up with a partial implementation that only applies to headings.

## Global Header
- Updated the global nav menus to open on hover instead of just on click.
- Removed icons from the global nav items and their children (in drop-downs). They were implemented inconsistently and used the AppBuilder supplied FontAwesome icons. This simplifies the presentation and prevents us from mixing multiple conflicting-styled icon systems.
- Removed the "About" navigation menu. It only contained a link to the "Version" view. Instead, I've taken the contents of the Version view and added a global footer that displays the same information.
- Simplified the Time Zone selector by removing the day and seconds. The hours are the most relevant part of the timestamp when it comes to timezones. The constant motion of the changing seconds felt like a bit of a distraction. Tweaked the styling within the dropdown as well.
- Replaced the current user's name with a circled monogram derived from their initials. Displays an icon fallback if first and last name values not present.
- ~I've partially removed the ability to define a custom background color of the navbar (`navbar_color`), it likely wouldn't translate well for users who currently utilizing it since the navbar was previously optimized for a dark background bar. Appears to be a bit of a superfluous feature. Is it something we can remove completely in order to ensure a clean presentation, or does the support need to be left in?~ (edit: retained this feature)
- I threw a little interaction easter egg in there—can you spot it?

**Before:**
image

**After:**
image

**Fallback icon when no first/last name present:**
Image 2020-09-29 at 1 35 21 PM

**Version info in global footer:**
image

## DAGs View (Home)
- Updated the table presentation with a cleaner look that should be a bit more pleasing to look at.
- Actions have been separated from links to better communicate their purpose. I do question wether the links should be here at all. They add a lot of visual noise and are certainly not intuitive for new users.

| Before | After |
|---|---|
| image  |  image |

## DAG Views
Most notable change across these views is the cleaned up run selection form fields. They now have a tidy presentation that is consistent across the views. I believe this further highlights the incongruence of the Task Instance navigation and why I've proposed #11089.

**Before:**

image

**After:**

image

## Flask AppBuilder CRUD Views

These views are automatically generated by AppBuilder, so there isn't access to markup/icon modifications (hence why they still use the Font Awesome icons). It is reliant on Bootstrap markup/classes, so I was able to update the colors to utilize the new theming.

image

## Misc.

Added all non-vendor stylesheets to the Webpack configuration. This outputs them with a fingerprinted filename and ensures the latest version isn't missed by the client due to caching.

### CODE REVIEW

**Comment by ryanahamilton** [comment]
Waiting until I've addressed any forthcoming feedback before cutting new screenshots for the Docs.

**Comment by ashb** [comment]
> First and Last name appear to be required fields—are there any cases where there might not be values and need a fallback?

When using various auth backends, either of these could be an empty string, which I think is "valid" according to the DB schema.

> Is it something we can remove completely in order to ensure a clean presentation, or does the support need to be left in?

This is often used by some users to give multiple environments a distinct colour ("green for staging, red for production" etc.)

File: airflow/www/extensions/init_views.py [hunk 1]
```diff
@@ -58,7 +58,6 @@ def init_appbuilder_views(app):
     appbuilder.add_view(views.PoolModelView, "Pools", category="Admin")
     appbuilder.add_view(views.VariableModelView, "Variables", category="Admin")
     appbuilder.add_view(views.XComModelView, "XComs", category="Admin")
-    appbuilder.add_view(views.VersionView, 'Version', category='About', category_icon='fa-th')
```
Comment (kaxil) [inline airflow/www/extensions/init_views.py:61]: Would be good to see the Screenshot of the footer where we can see the Airflow Version information

Comment (ryanahamilton) [reply airflow/www/extensions/init_views.py:61]: Ah, yes! Added this to the PR description too:
image

**Comment by ashb** [comment]
One comment on the DAG view:

The "running" dag is visually placed on the top right, but it is a property of which dag run is currently being selected. I'm not sure if this is a problem where you have it or not

**Comment by ryanahamilton** [comment]
@ashb:

1. Added a fallback icon when first/last name are not present. See updated PR description for screenshot.
2. I added the `navbar_color` capability back in.
3. I know what you mean about the "running" placement as it was moved from `graph.html` to `dag.html` template. It still only displays when on the Graph view—it didn't appear to cause any issues.

**Review by feluelle — COMMENTED** [review]
In the Graph View, what does "Runs" actually do? 
I mean isn't there always only one run displayed regardless of what you select in "Runs"?

And do we really need the "Update" button or can we update on changed setting?

**Comment by feluelle** [comment]
Nvm "Runs" adds the runs you can select from in "Run" correct?

**Comment by ryanahamilton** [comment]
> And do we really need the "Update" button or can we update on changed setting?

@feluelle I agree with your suggestion, though that's a refactor for another PR 😉

File: airflow/www/templates/airflow/code.html [hunk 2]
```diff
@@ -22,22 +22,22 @@
 {% block page_title %}{{ dag.dag_id }} - Code - Airflow{% endblock %}
 
 {% block content %}
-    {{ super() }}
-    <h2>{{ title }}</h2>
```
Comment (turbaszek) [inline airflow/www/templates/airflow/code.html:26]: @ryanahamilton have you used some liting? Those changes look unrelated but I'm ok with them if we can add linter to pre-commit 👍

Comment (mik-laj) [reply airflow/www/templates/airflow/code.html:26]: There is no good linter for Jinja2 😿.

**Comment by ausiddiqui** [comment]
This looks great! Had a few requests:

Is it possible to add a legend in the main _DAGs View (Home)_ for the Recent Tasks? There are many columns and colors and hard to keep track of. This might also be handled through a tooltip on the circle itself. 

When carrying out any action, such as "Mark as failed" or "Mark as success" a new page appears to confirm the action. The two action buttons are labeled "OK!" and "bail." in v1.10.10 (non RBAC UI). Can these labels be changed to just say "OK"/"Yes"/"Continue" and "Cancel"?

In much of the UI timestamps are exhibited like "09-16T13:45:08.842496+00:00" would there be a way to alter this format universally or in specific pages through config values. Execution times can be used as identifiers, but visually not sure the value of the milliseconds down to 6 digits of precision.

**Comment by WSHoekstra** [comment]
I love most of it, but there are a few things I'd like to see differently.
- I'll use the 'Search Dags' bar more than the 'Search by tags' search bar, so I'd like to see their positions reversed
- the existence of the 'Delete DAG' button makes it impossible to grant semi-technical users access to the UI (e.g. stakeholders who sometimes want to rerun relevant ETL's at will). Would it be possible to either hide the button altogether (after all - proper dag bag configuration should be done in code) or have an option in the admin settings to disable it there?

File: airflow/www/templates/airflow/code.html (hunk 2 above)
Comment (ryanahamilton) [reply airflow/www/templates/airflow/code.html:26]: Yeah, no linter. I'm just formatting as HTML files—indenting nested code blocks (w/ [2 spaces](https://github.com/apache/airflow/blob/master/.editorconfig#L48-L49)). There was very inconsistent formatting throughout the template files. Hopeful that introducing some consistency can at least suggest patterns to fellow contributors—even if they aren't enforceable w/ linting. It's kind of an obsessive behavior of mine driven by my Boy Scout days—leave your ~campsite~ code in better shape than you found it. 😃 

I will note that there is linting (esLint) in place for standalone JavaScript files that is currently not PR enforced (but did inform my formatting edits). We could/should enforce this, but the extent of the current offenses is so large that it should be addressed in a standalone PR.

**Review by ryw — APPROVED** [review]
Great work on this @ryanahamilton - with 2.0 alpha/beta testing we'll all be able to exercise the UI changes, and we can continue to refine the aesthetics - as long as everything works technically, I support merging this.

File: airflow/www/templates/airflow/dag_code.html [hunk 3]
```diff
@@ -22,34 +22,33 @@
...
-            });
-        }
-      });
-
-      // pygments generates the HTML so set wrap toggle via js
-      if ("{{ wrapped }}" == "True") {
-        toggleWrap();
-      };
-    </script>
+{% block tail_js %}
```
Comment (ashb) [inline airflow/www/templates/airflow/dag_code.html:32]: Any reason for different block in use here?

**Comment by kaxil** [comment]
> I love most of it, but there are a few things I'd like to see differently.
> 
> * I'll use the 'Search Dags' bar more than the 'Search by tags' search bar, so I'd like to see their positions reversed
> * the existence of the 'Delete DAG' button makes it impossible to grant semi-technical users access to the UI (e.g. stakeholders who sometimes want to rerun relevant ETL's at will). Would it be possible to either hide the button altogether (after all - proper dag bag configuration should be done in code) or have an option in the admin settings to disable it there?

The Delete DAG button won't be an issue coz from Airflow 2.0 RBAC UI would be the default UI and you can give DELETE permissions to only admins for example

File: airflow/www/templates/airflow/dag_code.html (hunk 3 above)
Comment (ryanahamilton) [reply airflow/www/templates/airflow/dag_code.html:32]: Just attempting to make the templates more consistent—not sure if I got them all. Presumably the `tail_js` comes after the `tail` as it's best practice for the scripts to be at the end of the document. Also assuming that the "js" in the block name designates the purpose.

**Review by ashb — APPROVED** [review]
Generally I prefer to keep "whitespace"/formatting only changes to their own PR/commit (so it'as easier to see in `git blame` what is formatting only, and what is functional.)

Not _strictly_ required in this case though.

File: airflow/www/templates/airflow/dag_code.html (hunk 3 above)
Comment (ashb) [reply airflow/www/templates/airflow/dag_code.html:32]: Gotcha :+1:

**Comment by ashb** [comment]
@ryanahamilton Look slike your version change isn't working. Tests are failing with:

```
E   jinja2.exceptions.UndefinedError: 'airflow_version' is undefined
```

**Comment by ashb** [comment]
Could you show a "full" page screenshot showing the global footer in place please?

Additionally: if the git version is not set (which I think is fairly common) we should just not show it in the HTML, rather than showing N/A

**Comment by ryanahamilton** [comment]
@ashb re: footer:

- Test failures have been resolved (there were a couple more than the one you spotted)
- I fixed issue where the airflow_version and git_version version weren't globally available (in FAB CRUD views).
- Updated the footer to only show to authenticated users as to not reveal version info
- Git version is no longer shown when there is no value
- Full screenshot of the footer:

Image 2020-10-01 at 11 13 07 AM

Updated Docs screenshots still to-do

File: airflow/www/extensions/init_jinja_globals.py [hunk 4]
```diff
@@ -42,6 +45,14 @@ def init_jinja_globals(app):
     expose_hostname = conf.getboolean('webserver', 'EXPOSE_HOSTNAME', fallback=True)
     hosstname = socket.getfqdn() if expose_hostname else 'redact'
 
+    try:
+        airflow_version = airflow.__version__
+    except Exception as e:  # pylint: disable=broad-except
+        airflow_version = None
+        logging.error(e)
+
+    git_version = get_airflow_git_version()
```
Comment (ryanahamilton) [inline airflow/www/extensions/init_jinja_globals.py:54]: @ashb Now that this is called for every page, [its internal logging](https://github.com/astronomer/airflow/blob/master/airflow/utils/platform.py#L60) when there is no git version has added quite a bit of noise when running the webserver—something we should consider removing?

Comment (kaxil) [reply airflow/www/extensions/init_jinja_globals.py:54]: Yeah or just change it to log.debug to be honest.

**Review by kaxil — APPROVED** [review]

**Comment by pcandoalmeida** [comment]
Looks awesome 🚀 Out of curiosity, if we wanted to introduce a `dark mode` in the 2.0 release (seems to be a trend these days) would it be relatively simple to extend from this work?

**Comment by ryanahamilton** [comment]
@pcandoalmeida thanks! I love the idea of allowing switching between light/dark modes too. Unfortunately the underlying design system framework (an old version of Bootstrap) is really getting in the way of that. We really need to move to a modern system where color values are tokenized in order to execute switching in a manner that remains sustainable as the UI continues to evolve. Now that the aesthetics are looking a bit more modern, I'm eager to work updating/upgrading these technologies as well.

**Comment by ryanahamilton** [comment]
@WSHoekstra moving the search box crossed my mind as well. I have some UX improvements in mind for that table header filtering that I'm looking to follow-up this PR with. I think that would be a good time for me to make that adjustment as well.

**Comment by ryanahamilton** [comment]
@ausiddiqui thanks!
- I really like the idea of adding a legend. As someone relatively new to Airflow, I found that to be one of the most confusing aspects of the UI. I think there are a lot of opportunities like this where we can improve the UX and onboarding process by doing a better job of teaching within the UI. Being a new feature in and of itself, I think it would extend a bit much out of the scope of this PR. This is definitely something I want to brainstorm more on—I think whatever a "legend" solution looks like should be a pattern that can be extended/implemented throughout the UI to create more consistent expectations for the user.
- That "bail" button appears to have already been addressed in #9499.
- I agree that those tooltips are not very user-friendly. I could see making that configurable, but even before that, it would be great if the formatting was consistent across the entire UI. Would be great to see an audit/update like that in a dedicated PR.

File: airflow/www/extensions/init_jinja_globals.py (hunk 4 above)
Comment (ryanahamilton) [reply airflow/www/extensions/init_jinja_globals.py:54]: Updated to `log.debug`

**Comment by himabindu07** [comment]
I have verified with this image quay.io/astronomer/ap-airflow-dev:2.0.0-buster-onbuild-22919

**Pull request merged**